core/src/main/scala/kafka/log/Log.scala

```diff
@@ -468,47 +467,50 @@ class Log(@volatile var dir: File,
     var lastOffset = -1L
     var sourceCodec: CompressionCodec = NoCompressionCodec
     var monotonic = true
-    var maxTimestamp = Record.NO_TIMESTAMP
+    var maxTimestamp = RecordBatch.NO_TIMESTAMP
     var offsetOfMaxTimestamp = -1L
-    for (entry <- records.shallowEntries.asScala) {
-      // update the first offset if on the first message
-      if(firstOffset < 0)
-        firstOffset = entry.offset
+
+    for (batch <- records.batches.asScala) {
+      // update the first offset if on the first message. For magic versions older than 2, we use the last offset
+      // to avoid the need to decompress the data (the last offset can be obtained directly from the wrapper message).
+      // For magic version 2, we can get the first offset directly from the batch header.
+      if (firstOffset < 0)
+        firstOffset = if (batch.magic >= RecordBatch.MAGIC_VALUE_V2) batch.baseOffset else batch.lastOffset
+
       // check that offsets are monotonically increasing
-      if(lastOffset >= entry.offset)
+      if (lastOffset >= batch.lastOffset)
         monotonic = false
       // update the last offset seen
-      lastOffset = entry.offset
-
-      val record = entry.record
+      lastOffset = batch.lastOffset
 
       // Check if the message sizes are valid.
-      val messageSize = entry.sizeInBytes
-      if(messageSize > config.maxMessageSize) {
+      val batchSize = batch.sizeInBytes
+      if (batchSize > config.maxMessageSize) {
         BrokerTopicStats.getBrokerTopicStats(topicPartition.topic).bytesRejectedRate.mark(records.sizeInBytes)
         BrokerTopicStats.getBrokerAllTopicsStats.bytesRejectedRate.mark(records.sizeInBytes)
-        throw new RecordTooLargeException("Message size is %d bytes which exceeds the maximum configured message size of %d."
-          .format(messageSize, config.maxMessageSize))
+        throw new RecordTooLargeException(s"Message size is $batchSize bytes which exceeds the maximum configured " +
+          s"message size of ${config.maxMessageSize}.")
```

Comment:
This is a good point. The current usage of `RecordBatchTooLargeException` is to indicate when a batch exceeds the log segment size, but this only really made sense for the old format without compression. It might be worthwhile considering in a follow-up whether we have need to reuse this error for other purposes. For now, I'll update the message.